tuf_conformance/test_updater_delegation_graphs.py

```diff
@@ -0,0 +1,196 @@
+        delegations=[
+            DelegationTester("targets", "A"),
+            DelegationTester("targets", "B"),
+            DelegationTester("A", "C", paths=["*.py"]),
+            DelegationTester("C", "D"),
+        ],
+        visited_order=["A", "B"],
+    ),
+    "cyclic-graph": DelegationsTestCase(
+        delegations=[
+            DelegationTester("targets", "A"),
+            DelegationTester("targets", "B"),
+            DelegationTester("B", "C"),
+            DelegationTester("C", "D"),
+            DelegationTester("D", "B"),
+        ],
+        visited_order=["A", "B", "C", "D"],
+    ),
+    "two-roles-delegating-to-a-third": DelegationsTestCase(
+        delegations=[
+            DelegationTester("targets", "A"),
+            DelegationTester("targets", "B"),
+            DelegationTester("B", "C"),
+            DelegationTester("A", "C"),
+        ],
+        # Under all same conditions, 'C' is reached through 'A' first"
+        visited_order=["A", "C", "B"],
+    ),
+    "two-roles-delegating-to-a-third-different-paths": DelegationsTestCase(
+        delegations=[
+            DelegationTester("targets", "A"),
+            DelegationTester("targets", "B"),
+            DelegationTester("B", "C"),
+            DelegationTester("A", "C", paths=["*.py"]),
+        ],
+        # 'C' is reached through 'B' since 'A' does not delegate a matching pattern"
+        visited_order=["A", "B", "C"],
+    ),
+}
+
+
+graph_ids = graphs.keys()
+graph_cases = graphs.values()
+
+
+def init_repo(repo: RepositorySimulator, test_case: DelegationsTestCase) -> None:
+    spec_version = ".".join(SPECIFICATION_VERSION)
+    for d in test_case.delegations:
+        if d.rolename in repo.mds:
+            targets = repo.mds[d.rolename].signed
+        else:
+            targets = Targets(1, spec_version, repo.safe_expiry, {}, None)
+        # unpack 'd' but skip "delegator"
+        role = DelegatedRole(*astuple(d)[1:])
+        repo.add_delegation(d.delegator, role, targets)
+
+    for target in test_case.target_files:
+        repo.add_artifact(*astuple(target))
+
+    repo.update_snapshot()
+
+
+@pytest.mark.parametrize("graphs", graph_cases, ids=graph_ids)
+def test_graph_traversal(
+    client: ClientRunner, server: SimulatorServer, graphs: DelegationsTestCase
+) -> None:
+    """Test that delegated roles are traversed in the order of appearance
+    in the delegator's metadata, using pre-order depth-first search"""
+
+    if "clients/go-tuf/go-tuf" in client._cmd[0]:
+        pytest.skip("skip for flakiness")
+
+    exp_calls = [(role, 1) for role in graphs.visited_order]
+
+    init_data, repo = server.new_test(client.test_name)
+
+    assert client.init_client(init_data) == 0
+    init_repo(repo, graphs)
+
+    # Call explicitly refresh to simplify the expected_calls list
+    assert client.refresh(init_data) == 0
+    repo.metadata_statistics.clear()
+    client.download_target(init_data, "missingpath")
```

Comment:
could still check for return value -- I think especially the expected failures are useful for the reader:  then we don't necessarily need to call that out in a comment

Reply:
Sorry, should be there now with https://github.com/theupdateframework/tuf-conformance/pull/115/commits/c9b930dfdb1bbddf3973c00985b2fba58dfcf81a